java-checks-testkit/src/main/java/org/sonar/java/checks/verifier/TestCheckRegistrarContext.java

```diff
@@ -52,16 +54,34 @@ public void registerTestChecks(String repositoryKey, Collection<?> javaCheckClas
 
   @Override
   public void registerMainSharedCheck(JavaCheck check, Collection<RuleKey> ruleKeys) {
+    registerMainHook(check);
+    mainRuleKeys.addAll(ruleKeys);
+  }
+
+  private void registerMainHook(JavaCheck check) {
     mainCheckClasses.add(check.getClass());
     mainCheckInstances.add(check);
-    mainRuleKeys.addAll(ruleKeys);
   }
 
   @Override
   public void registerTestSharedCheck(JavaCheck check, Collection<RuleKey> ruleKeys) {
+    registerTestHook(check);
+    testRuleKeys.addAll(ruleKeys);
+  }
+
+  private void registerTestHook(JavaCheck check) {
     testCheckClasses.add(check.getClass());
     testCheckInstances.add(check);
-    testRuleKeys.addAll(ruleKeys);
+  }
+
+  @Override
+  public void registerCustomFileScanner(RuleScope ruleScope, JavaFileScanner scanner) {
+    if (ruleScope != RuleScope.TEST) {
+      registerMainHook(scanner);
+    }
+    if (ruleScope != RuleScope.MAIN) {
+      registerTestHook(scanner);
+    }
```

Comment:
If we change the implemetation in `SonarComponents`, we might as well change it here